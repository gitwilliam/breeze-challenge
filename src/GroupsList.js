import React, { Component } from 'react'
import { Table } from 'semantic-ui-react'
import _ from 'lodash'

class GroupsList extends Component {
    constructor(props) {
        super(props);
        this.state = { column: null, data: [], direction: null };
    }

    componentDidMount() {
        // fetch("http://localhost:8000/api/groups")
        //   .then(response => response.json())
        //   .then(data => this.setState({ data: data.data }));
        fetch("http://localhost:8000/api/groups/members")
          .then(response => response.json())
          .then(data => this.setState({ data: data.data }));
    }

    handleSort = (clickedColumn) => () => {
      const { column, data, direction } = this.state

      if (column !== clickedColumn) {
        this.setState({
          column: clickedColumn,
          data: _.sortBy(data, [clickedColumn]),
          direction: 'ascending',
        })

        return
      }

      this.setState({
        data: data.reverse(),
        direction: direction === 'ascending' ? 'descending' : 'ascending',
      })
    }

    render() {
        const { column, data, direction } = this.state

        return (
            <Table celled sortable structured>
              <Table.Header>
                <Table.Row>
                <Table.HeaderCell sorted={column === 'group_name' ? direction : null}
                  onClick={this.handleSort('group_name')}>Group Name</Table.HeaderCell>
                <Table.HeaderCell sorted={column === 'first_name' ? direction : null}
                  onClick={this.handleSort('first_name')}>First Name</Table.HeaderCell>
                <Table.HeaderCell sorted={column === 'last_name' ? direction : null}
                  onClick={this.handleSort('last_name')}>Last Name</Table.HeaderCell>
                <Table.HeaderCell sorted={column === 'email_address' ? direction : null}
                  onClick={this.handleSort('email_address')}>Email</Table.HeaderCell>
                </Table.Row>
              </Table.Header>

              <Table.Body>

              {
                  data.map((group, index) => {
                    if (group.members.length === 0) {
                      return (
                        <Table.Row key={index}>
                          <Table.Cell rowSpan={group.members.length}>{ group.group_name }</Table.Cell>
                          <Table.Cell></Table.Cell>
                          <Table.Cell></Table.Cell>
                          <Table.Cell></Table.Cell>
                        </Table.Row>
                      )
                    }

                    return (
                      <>
                        <Table.Row key={index}>
                          <Table.Cell rowSpan={group.members.length}>{ group.group_name }</Table.Cell>
                          <Table.Cell>{ group.members[0].first_name }</Table.Cell>
                          <Table.Cell>{ group.members[0].last_name }</Table.Cell>
                          <Table.Cell>{ group.members[0].email_address }</Table.Cell>
                        </Table.Row>
                          {
                            group.members.slice(1).map((member) => {
                              return (
                              <Table.Row>
                                <Table.Cell>{ member.first_name }</Table.Cell>
                                <Table.Cell>{ member.last_name }</Table.Cell>
                                <Table.Cell>{ member.email_address }</Table.Cell>
                              </Table.Row>
                              );
                            })
                          }
                        </>
                    );
                  })
              }

              </Table.Body>
            </Table>
    );
}

}

export default GroupsList
