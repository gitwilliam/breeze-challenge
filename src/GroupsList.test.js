import React from 'react';
import { shallow } from 'enzyme';
import GroupsList from './GroupsList';

let wrapper, data;

describe('<GroupsList />', () => {

    beforeAll(() => {
        wrapper = shallow(<GroupsList />)
        data = [{
            "id": 132,
            "group_name": "Elders"
        }]

        wrapper.setState({'column': null, 'data' : data, 'direction': null });
    });

    test('should match the snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    });
})
