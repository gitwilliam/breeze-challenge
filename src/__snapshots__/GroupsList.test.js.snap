// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GroupsList /> should match the snapshot 1`] = `
<Tab
  grid={
    Object {
      "paneWidth": 12,
      "tabWidth": 4,
    }
  }
  menu={
    Object {
      "attached": true,
      "tabular": true,
    }
  }
  panes={
    Array [
      Object {
        "menuItem": "Elders",
        "pane": <TabPane
          active={true}
          as={[Function]}
        >
          <PeopleList
            group_id={132}
          />
        </TabPane>,
      },
    ]
  }
  renderActiveOnly={false}
/>
`;
